### PR TITLE
Add & document msg_capacity, msg_reserve

### PR DESCRIPTION
**Pull request by EvanBalster**
This PR creates and documents two new functions dealing with message capacity — the maximum length to which a message can grow without invoking the memory allocator.

`nng_msg_capacity` returns the current capacity based on the underlying body chunk (not counting headroom)
`nng_msg_reserve` will expand capacity to be greater than or equal to the given value, but will in no case decrease message length or headroom.  (Under the hood, it forwards to `nni_chunk_grow`)

Please double check that my changes to the documentation were done correctly.

### CODE REVIEW

File: src/core/message.c [hunk 1]
```diff
@@ -456,6 +456,18 @@ nni_msg_realloc(nni_msg *m, size_t sz)
 	return (0);
 }
 
+int
+nni_msg_reserve(nni_msg *m, size_t capacity)
+{
+	return nni_chunk_grow(&m->m_body, capacity, 0);
```
Comment (gdamore) [inline src/core/message.c:462]: style nit:

we parenthesize the return value always.

File: src/core/message.c [hunk 2]
```diff
@@ -456,6 +456,18 @@ nni_msg_realloc(nni_msg *m, size_t sz)
...
+}
+
+size_t
+nni_msg_capacity(nni_msg *m)
+{
+	return (size_t) ((m->m_body.ch_buf + m->m_body.ch_cap) - m->m_body.ch_ptr);
```
Comment (gdamore) [inline src/core/message.c:468]: ditto the style -- in this case the parenthesis should include the cast.

e.g.

return ((size_t) (....))

File: docs/man/nng_msg_capacity.3.adoc [hunk 3]
```diff
@@ -0,0 +1,46 @@
+= nng_msg_capacity(3)
```
Comment (gdamore) [inline docs/man/nng_msg_capacity.3.adoc:1]: Can you add a few more cross references back to these pages?

Specifically in nng_msg.5.adoc, and nng_msg_len.3.adoc and nng_msg_realloc.3.adoc?

**Comment by gdamore** [comment]
Looks pretty good, except for a few more doc items and a style nits.

**Comment by neachdainn** [comment]
Do you want to always grow the chunk with `nng_msg_reserve`? It might make sense to add a check against the current capacity.

File: docs/man/nng_msg_capacity.3.adoc (hunk 3 above)
Comment (EvanBalster) [reply docs/man/nng_msg_capacity.3.adoc:1]: There's already a cross-reference in realloc (possibly I missed the list at the bottom, though).  I'll attend to the others.

**Comment by gdamore** [comment]
> Do you want to always grow the chunk with `nng_msg_reserve`? It might make sense to add a check against the current capacity.

This isn't necessary -- nni_chunk_grow already has these checks and won't actually change the message if the capacity is already as large as (or larger than) the requested capacity.

**Comment by EvanBalster** [comment]
Style nits addressed, cross-references added.

> Do you want to always grow the chunk with `nng_msg_reserve`? It might make sense to add a check against the current capacity.

I think this is the behavior of `nni_chunk_grow`.

**Comment by gdamore** [comment]
Prolly we should have put some test case coverage on this.  However, tests/message.c really needs to be converted over to NUTS so I don't particularly feel like asking you to fix that.

**Pull request merged**